src/vtl.ts

```diff
@@ -418,4 +437,78 @@ export class VTL {
     const __exhaustive: never = node;
     return __exhaustive;
   }
+
+  private mapForEachArgs(call: CallExpr) {
+    const fn = call.args.callbackfn as FunctionExpr;
+    return fn.parameters.map((p) => p.name);
+  }
+
+  private renderMapOrForEach(
+    call: CallExpr,
+    list: string,
+    // Should start with $
+    returnVariable?: string,
+    index?: string,
+    array?: string
+  ) {
+    if (index) {
+      this.add(`#set($${index} = $foreach.index)`);
+    }
+    if (array) {
+      this.add(`#set($${array} = ${list})`);
+    }
+
+    const fn = call.args.callbackfn as FunctionExpr;
+
+    const tmp = returnVariable ? returnVariable : this.newLocalVarName();
+
+    for (const stmt of fn.body.statements) {
+      this.eval(stmt, tmp);
+    }
+
+    return tmp;
+  }
+
+  /**
+   * Recursively dives into map operations until a non-map or a map with arr paremeter is found.
+   * evaluates the expression after the last map.
+   *
+   * @return [firstVariable, list variable, render function]
+   */
+  private divePropAccessMapOperations(
```

Comment:
The name of this function doesn't immediately tell me what it does.

Reply:
uhh, yeah, will think of a new name...